sympy/functions/elementary/trigonometric.py

```diff
@@ -107,20 +107,21 @@ def _peeloff_pi(arg):
     (x + pi*y + pi/6, pi/2)
 
     """
+    pi_coeff = S.Zero
+    rest_terms = []
     for a in Add.make_args(arg):
-        if a is S.Pi:
-            K = S.One
-            break
-        elif a.is_Mul:
-            K, p = a.as_two_terms()
-            if p is S.Pi and K.is_Rational:
-                break
-    else:
+        K = a.coeff(S.Pi)
+        if K and K.is_rational:
+            pi_coeff += K
+        else:
+            rest_terms.append(a)
+
+    if pi_coeff is S.Zero:
         return arg, S.Zero
 
-    m1 = (K % S.Half)*S.Pi
-    m2 = K*S.Pi - m1
-    return arg - m2, m2
+    m1 = (pi_coeff % S.Half)*S.Pi
+    m2 = pi_coeff*S.Pi - m1
+    return Add(*rest_terms) + m1, m2
```

Comment:
This calls Add.flatten twice. You can make it a single call with `Add(*(rest_terms + [m1]))`

Reply:
Anything else?